Source/Scene/ModelOutlineLoader.js

```diff
@@ -37,10 +80,20 @@ ModelOutlineLoader.hasExtension = function (model) {
  * @private
  */
 ModelOutlineLoader.outlinePrimitives = function (model) {
-  if (!ModelOutlineLoader.hasExtension(model)) {
+  if (
+    !ModelOutlineLoader.hasExtension(model) &&
+    !ModelOutlineLoader.shouldGenerateOutlines(model)
+  ) {
     return;
   }
 
+  if (ModelOutlineLoader.shouldGenerateOutlines(model)) {
```

Comment:
The short answer is yes, most JS engines will cache the output of most functions if they get called multiple times with the same parameters. It doesn't _always_ do this though, optimising compilers generally only bother if it's worth the effort— like if a function is being called many, many times. 

The exact circumstances under which a particular bit of code will be optimised or not are hard to pin down, because the rules are complicated and subtle. As a general rule of thumb, I like to write code that is easy for compilers to optimise, as if it's not going to get optimised. So yes, even though the compiler would probably cache the result of this function if it got called a bunch of times, I should cache the result, you're absolutely right.

I couldn't find a good article on this exact scenario but https://ponyfoo.com/articles/an-introduction-to-speculative-optimization-in-v8 and https://mathiasbynens.be/notes/shapes-ics are some long reads that explain some of the general principles at play.

Reply:
Oh EXCEPT I just went to implement caching for it and it's kind of a pain because ModelOutlineLoader is a static class and doesn't store state per model. so never mind

Reply:
Ah right, what if it was just a `var shouldGenerateOutlinesForModel = shouldGenerateOutlines(model)`? 🤔 and use that boolean instead of calling the function twice? 
Though now that you added the break statement it's probably not such a bad thing to leave this out since in theory the `shouldGenerateOutlines` function shouldn't take as long to run :)